src/models/history/tree-manager.ts

```diff
@@ -504,12 +601,97 @@ export const TreeManager = types
     // and print a warning to the console.
     // One way might be using unique history_entry_ids that we mark as closed
     // after the finish call.
+    self.currentHistoryIndex = newHistoryIndex;
   }),
 
+  // Unlike currentHistoryIndex, the historyIndex here is index into
+  // history entry array. So a historyIndex of 0 means 
+  // "give me the first history entry"
   getHistoryEntry: (historyIndex: number) => {
     return self.document.history.at(historyIndex);
   }
 
 }));
 
 export interface TreeManagerType extends Instance<typeof TreeManager> {}
+
+const historyEntryConverter = {
+  toFirestore: (entry: HistoryEntrySnapshot) => {
+    throw new Error(
+      "We can't convert a raw HistoryEntry to firestore because we need the index from its parent collection");
+  },
+  fromFirestore: (doc: firebase.firestore.QueryDocumentSnapshot): HistoryEntrySnapshot => {
+    const { entry } = doc.data();
+    return JSON.parse(entry);
+  }
+};
+
+interface IPrepareFirestoreHistoryInfoArgs {
+  userContext?: IUserContext;
+  mainDocument?: IMainDocument;
+  firestore?: Firestore;
+}
+
+async function prepareFirestoreHistoryInfo(
+    {userContext, mainDocument, firestore}: IPrepareFirestoreHistoryInfoArgs): Promise<IFirestoreHistoryInfo> {
+
+  if (!userContext || !mainDocument || !firestore || !userContext.uid) {
+    console.error("cannot record history entry because environment is not valid",
+      { userContext, mainDocument, firestore });
+    throw new Error("cannot record history entry because environment is not valid");
+  }
+    
+  const documentPath = getDocumentPath(userContext.uid, mainDocument.key, userContext.network);
+  const documentRef = firestore.doc(documentPath);
+  const docSnapshot = await documentRef.get();
+  // create a document if necessary
+  if (!docSnapshot.exists) {
+    const validateCommentableDocument = 
+      getFirebaseFunction<ICommentableDocumentParams>("validateCommentableDocument_v1");
+
+
+    // FIXME-HISTORY: rename this function to validateFirestoreDocumentMetadata_v1
+    await validateCommentableDocument({context: userContext, document: mainDocument.metadata});
+  }
+
+  const lastHistoryEntry = await getLastHistoryEntry(firestore, documentPath);
+
+  return {
+    documentPath, 
+    // We start with -1 so if there is no last entry the next entry will get an index of 0
+    // 0 is a valid index so ?? must be used instead of ||
+    lastEntryIndex: lastHistoryEntry?.index ?? -1, 
+    // We use null here so this is a valid Firestore property value
+    lastEntryId: lastHistoryEntry?.id || null
+  };
+}
+
+function getDocumentPath(userId: string, documentKey: string, network?: string) {
```

Comment:
This function looks like it should be in a more general utils file to me.

Reply:
Yeah I agree, but I'm stuck on figuring out what name to give it. `getDocumentPath` is too generic. It most closely matches `useCommentableDocumentPath`. But this isn't just a commentable document anymore. So far this specific function can't be used anywhere else since it only works with user documents not curriculum documents. Whereas useCommentableDocumentPath work with both.

So I think I'll keep it in here for now. I'll propose a few names for this type of document in slack to see if we can come with name for these history and comment parent documents.